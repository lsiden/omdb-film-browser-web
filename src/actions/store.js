import { createStore, applyMiddleware, compose } from "redux"
import thunk from "redux-thunk"

import Actions from "actions/types"

const initialState = {
  view: Actions.VIEW_FILM_LIST,
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.VIEW_FILM_LIST:
    case Actions.VIEW_FILM_DETAIL:
    case Actions.UPDATE_FILMS:
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
      return {
        ...state,
        ...data,
      }
    case Actions.APPEND_FILMS:
      return {
        ...state,
        ...data,
        films: [...state.films, ...data.films],
      }
    default:
      return state
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
