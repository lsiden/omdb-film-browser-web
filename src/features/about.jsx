import React from 'react'
import ReactMarkdown from 'react-markdown'

import CloseButton from 'components/close-button'
import { headerStyle } from 'style'
import { BANNER_TITLE, ESC_KEY } from 'omdb_constants'

const OMDB_API_URL = 'https://www.omdbapi.com/'
const SOURCE_CODE_URL = 'https://github.com/lsiden/omdb-film-browser-web'
const WESTSIDE_CONSULTING_URL = 'http://westsideconsultingllc.com'

const source = `
## Written By
Lawrence Siden

[Westside Consulting LLC](${WESTSIDE_CONSULTING_URL})

[westsideconsultingllc@gmail.com](mailto:westsideconsultingllc@gmail.com)

## Credits
[Open Movie Database API](${OMDB_API_URL})

## Source Code
${SOURCE_CODE_URL}
`

const aboutBannerStyle = {
  ...headerStyle,
  display: 'flex',
  justifyContent: 'space-between',
}

class _About extends React.Component {
  static onKeyDown(ev) {
    if (ev.keyCode === ESC_KEY) {
      window.history.back()
    }
  }

  componentDidMount() {
    document.addEventListener('keydown', _About.onKeyDown)
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', _About.onKeyDown)
  }

  render() {
    return (
      <div style={{ outline: 'none' }}>
        <header style={aboutBannerStyle}>
          <h1 style={{ fontSize: 28 }}>{`About ${BANNER_TITLE}`}</h1>
          <CloseButton buttonStyle={{ color: 'white' }} />
        </header>
        <div style={{ marginLeft: 20 }}>
          <ReactMarkdown source={source} />
        </div>
      </div>
    )
  }
}

export default _About
