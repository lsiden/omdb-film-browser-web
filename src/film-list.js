import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import FilmTitle from 'film-title'

const ulStyle = {
	listStyleType: 'none',
	lineHeight: 1.5,
}

export const FilmList = ({films}) => {
	const msgStyle = {
		fontSize: '14pt',
		color: '#3266ba',
		margin: '1em'
	}
	if (films.length > 0) {
		return (
			<ul style={ulStyle}>{
				films.map(filmSummary => (
					<FilmTitle
						key={filmSummary.imdbID}
						filmSummary={filmSummary}
					/>
				))
			}
			</ul>
		)
	} else {
		return <div style={msgStyle}>There are no films that match your query yet.</div>
	}
}

FilmList.propTypes = {
	films: PropTypes.arrayOf(PropTypes.object),
}
FilmList.defaultProps = {
	films: [],
}

export default connect(
	state => ({
		films: state.films,
	}),
)(FilmList)
