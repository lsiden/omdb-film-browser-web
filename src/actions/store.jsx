import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import Actions from 'actions/types'
import { updateNewFilms, appendFilms } from 'actions/films'

const initialState = {
  view: Actions.VIEW_FILM_LIST,
  isFetching: false,
  films: [],
  totalResults: 0,
  query: '',
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
    case Actions.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case Actions.UPDATE_FILMS:
      return updateNewFilms(state, data)
    case Actions.APPEND_FILMS:
      return appendFilms(state, data)
    default:
      return state
  }
}

const composeEnhancers = (
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
  ) || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
