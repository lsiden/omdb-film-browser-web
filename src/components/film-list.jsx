import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import FilmTitle from 'components/film-title'
import { getFilms } from 'store'

const ulStyle = {
  listStyleType: 'none',
  lineHeight: 1.5,
  marginTop: 20,
}

export const _FilmList = ({ films }) => (
  <ul style={ulStyle}>
    {films.map((film) => (
      <FilmTitle key={film.imdbID} filmSummary={film} />
    ))}
  </ul>
)

_FilmList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object).isRequired,
}

export default connect(
  (state) => ({
    films: getFilms(state),
  }),
)(_FilmList)
