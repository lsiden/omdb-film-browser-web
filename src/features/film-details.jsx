import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Spinner from 'respin'

import CloseButton from 'components/close-button'
import NavButton from 'components/nav-button'
import { scrollToTop } from 'components/scroll'
import { fetchFilmDetails } from 'actions/remote'
import { headerStyle as bannerHeaderStyle } from 'style'
import { ESC_KEY, BANNER_TITLE, FETCH_TIMEOUT } from 'constants'

const titleStyle = {
  marginBottom: '5pt',
}
const detailsStyle = {
  marginTop: 20,
  fontSize: '14pt',
}
const topButtonStyle = {
  marginBottom: 20,
}
const spinnerWrapperStyle = {
  height: '100vh',
  minHeight: '100vh',
  display: 'flex',
  justifyContent: 'center',
}

const itemExists = (item) => item && item !== 'N/A'
const imdbUrl = (imdbID) => `https://www.imdb.com/title/${imdbID}`

const titleBannerStyle = {
  ...bannerHeaderStyle,
  display: 'flex',
  justifyContent: 'space-between',
}

const TitleBanner = ({ title }) => (
  <header style={titleBannerStyle}>
    <div style={{ lineHeight: 1 }}>
      <h1 style={{ fontSize: 24 }}>{BANNER_TITLE}</h1>
      <br />
      <h1 style={{ ...titleStyle, marginTop: 0 }}>{title}</h1>
    </div>
    <CloseButton buttonStyle={{ color: 'white' }} />
  </header>
)
TitleBanner.propTypes = {
  title: PropTypes.string.isRequired,
}

export class FilmDetails_ extends React.Component {
  static keyDownListener(ev) {
    if (ev.keyCode === ESC_KEY) {
      window.location.href = '/'
    }
  }

  constructor(props) {
    const { imdbID } = props
    super(props)
    this.imdbUrl = `https://www.imdb.com/title/${imdbID}`
  }

  componentDidMount() {
    document.addEventListener('keydown', FilmDetails_.keyDownListener)
    setTimeout(() => {
      if (!this.props.filmDetails) {
        const { dispatchFetchFilmDetails, imdbID } = this.props
        dispatchFetchFilmDetails(imdbID)
      }
    }, FETCH_TIMEOUT)
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', FilmDetails_.keyDownListener)
  }

  renderTitle() {
    const { filmDetails } = this.props
    return <TitleBanner title={filmDetails.Title} />
  }

  renderDetails() {
    const { filmDetails } = this.props
    const FilmDetail = ({ name, label }) => (
      <li>{label || name}: {filmDetails[name]}</li>
    )

    return (
      <ul style={detailsStyle}>
        <FilmDetail name="Year" />
        <FilmDetail name="Director" />
        <FilmDetail name="Writer" />
        <FilmDetail name="Actors" label="Cast" />
        <FilmDetail name="Language" />
        {itemExists(filmDetails.Awards) && <FilmDetail name="Awards" />}
        <FilmDetail name="Runtime" label="Length" />
        <FilmDetail name="imdbRating" label="IMDB Rating" />
        <FilmDetail name="BoxOffice" />
        {
          itemExists(filmDetails.Website)
          && <FilmDetail key="Website" label="Official Website" />
        }
        <li>
          <a href={imdbUrl(filmDetails.imdbID)}>IMDB page</a>
        </li>
      </ul>
    )
  }

  static renderScrollToTopButton() {
    return (
      <NavButton
        style={topButtonStyle}
        onClick={scrollToTop}
        title="Scroll to top of page"
      >
        Top
      </NavButton>
    )
  }

  render() {
    const { filmDetails } = this.props

    if (!filmDetails) {
      return (
        <div style={spinnerWrapperStyle}>
          <Spinner size={64} />
        </div>
      )
    } else {
      return (
        <div>
          {this.renderTitle()}
          <div style={{ marginLeft: 20 }}>
            <img
              src={filmDetails.Poster}
              alt="poster"
              style={{ marginTop: 10 }}
            />
            {this.renderDetails()}
            {FilmDetails_.renderScrollToTopButton()}
          </div>
        </div>
      )
    }
    return null
  }
}

FilmDetails_.propTypes = {
  imdbID: PropTypes.string.isRequired,
  filmDetails: PropTypes.object,
  dispatchFetchFilmDetails: PropTypes.func.isRequired,
}
FilmDetails_.defaultProps = {
  filmDetails: null,
}

export default connect(
  (state) => ({
    filmDetails: state.filmDetails,
  }),
  (dispatch) => ({
    dispatchFetchFilmDetails: (imdbID) => dispatch(fetchFilmDetails(imdbID)),
  }),
)(FilmDetails_)
