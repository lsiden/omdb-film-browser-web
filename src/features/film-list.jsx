import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Spinner from 'respin'

import FilmTitle from 'components/film-title'
import MoreButton from 'components/more-button'
import { scrollToTop, scrollToBottom } from 'components/scroll'
import NavButton from 'components/nav-button'
import { setQuery } from 'actions'
import { queryFetch } from 'actions/remote'
import { TITLE_COLOR } from 'omdb_constants'
import Banner from 'components/banner'

const ulStyle = {
  listStyleType: 'none',
  lineHeight: 1.5,
  marginTop: 20,
}
const msgStyle = {
  fontSize: '14pt',
  color: TITLE_COLOR,
  margin: '1em',
}
const bottomRowStyle = {
  marginLeft: 40,
  display: 'flex',
  justifyContent: 'flex-start',
  alignItems: 'baseline',
}
const topButtonStyle = {
  marginLeft: 40,
  fontSize: 16,
  color: 'darkgrey',
}
const spinnerStyle = {
  margin: 50,
}

export class _FilmList extends React.Component {
  constructor(props) {
    const { query, dispatchSetQuery } = props
    super(props)
    dispatchSetQuery(query)
  }

  componentDidUpdate() {
    scrollToBottom()
  }

  componentDidCatch(err, errInfo) {
    console.error(err, errInfo)
  }

  renderInner() {
    const { query, films, totalResults, isFetching } = this.props

    if (isFetching) {
      return <Spinner style={spinnerStyle} />
    }

    if (query.length === 0) {
      return <div style={msgStyle}> Search for a title. </div>
    }

    if (query.length > 0 && films.length === 0) {
      return <div style={msgStyle}> There are no films that match your query. </div>
    }

    return (
      <div>
        <ul style={ulStyle}>
          {films.map((filmSummary) => (
            <FilmTitle key={filmSummary.imdbID} filmSummary={filmSummary} />
          ))}
        </ul>
        <div style={bottomRowStyle}>
          {films.length < totalResults && <MoreButton />}
          { films && films.length && (
          <NavButton
            onClick={scrollToTop}
            style={topButtonStyle}
            title="Scroll to top of page"
          >top
          </NavButton>
          )}
        </div>
      </div>
    )
  }

  render() {
    return (
      <div>
        <Banner />
        {this.renderInner()}
      </div>
    )
  }
}

_FilmList.propTypes = {
  query: PropTypes.string,
  films: PropTypes.arrayOf(PropTypes.object),
  totalResults: PropTypes.number,
  dispatchSetQuery: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
}

_FilmList.defaultProps = {
  query: '',
  films: [],
  totalResults: 0,
}

export default connect(
  (state, ownProps) => ({
    query: ownProps.query,
    films: state.films,
    totalResults: parseInt(state.totalResults, 10) || 0,
    isFetching: state.isFetching || false,
  }),
  (dispatch) => ({
    dispatchSetQuery: (query) => {
      dispatch(setQuery(query))
      dispatch(queryFetch(query))
    },
  }),
)(_FilmList)
