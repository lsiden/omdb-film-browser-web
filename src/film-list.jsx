import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import FilmTitle from 'film-title'
import MoreButton from 'components/more-button'
import { scrollToTop, scrollToBottom } from 'components/scroll'
import ButtonLink from 'components/button-link'
import QueryForm from 'query-form'
import { headerStyle, titleStyle } from 'style'
import { setQuery } from 'actions'
import { BANNER_TITLE, TITLE_COLOR } from './constants'

const Banner = () => (
  <header style={headerStyle}>
    <div style={{ display: 'flex', justifyContent: 'space-between' }}>
      <h1 style={titleStyle}>{BANNER_TITLE}</h1>
      <a href="/about" style={{ color: 'white' }}>
        about
      </a>
    </div>
    <QueryForm />
  </header>
)

const ulStyle = {
  listStyleType: 'none',
  lineHeight: 1.5,
  marginTop: 20,
}
const msgStyle = {
  fontSize: '14pt',
  color: TITLE_COLOR,
  margin: '1em',
}

const bottomRowStyle = {
  marginLeft: 40,
  display: 'flex',
  justifyContent: 'flex-start',
  alignItems: 'baseline',
}

const topButtonStyle = {
  marginLeft: 40,
  fontSize: 16,
  color: 'darkgrey',
}

const renderFilmList = ({ films = [], totalResults = 0 }) => (
  <div>
    <Banner />
    {films.length === 0 ? (
      <div>
        <div style={msgStyle}>
          There are no films that match your query.
        </div>
      </div>
    ) : (
      <div>
        <ul style={ulStyle}>
          {films.map((filmSummary) => (
            <FilmTitle key={filmSummary.imdbID} filmSummary={filmSummary} />
          ))}
        </ul>
        <div style={bottomRowStyle}>
          {films.length < totalResults && <MoreButton />}
          <ButtonLink
            onClick={scrollToTop}
            style={topButtonStyle}
            title="Scroll to top of page"
          >
            top
          </ButtonLink>
        </div>
      </div>
    )}
  </div>
)

renderFilmList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object),
  totalResults: PropTypes.number,
}

renderFilmList.defaultProps = {
  films: [],
  totalResults: 0,
}

class FilmList extends React.Component {
  componentDidUpdate() {
    scrollToBottom()
  }

  /* eslint-disable-next-line class-methods-use-this */
  componentDidCatch(err, errInfo) {
    console.error(err, errInfo)
  }

  render() {
    return renderFilmList(this.props)
  }
}

const ConnectedFilmList = connect((state) => ({
  films: state.films,
  totalResults: state.totalResults,
}))(FilmList)

const RoutedFilmList = ({ match, dispatchSetQuery }) => {
  dispatchSetQuery(match.params.query)
  return <ConnectedFilmList />
}

RoutedFilmList.propTypes = {
  match: PropTypes.object.isRequired,
  dispatchSetQuery: PropTypes.func.isRequired,
}

export default connect(null, (dispatch) => ({
  dispatchSetQuery: (query) => dispatch(setQuery(query)),
}))(RoutedFilmList)
