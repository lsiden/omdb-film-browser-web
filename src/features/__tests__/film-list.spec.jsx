import React from 'react'
import { create } from 'react-test-renderer'
import { StaticRouter } from 'react-router'
import { Provider } from 'react-redux'

import { _FilmList } from 'features/film-list'
import FilmTitle from 'components/film-title'
import store from 'actions/store'

const films = require('./films.json').Search

it('renders list of titles', () => {
  const dispatchSetQuery = jest.fn()
  const testInstance = create(
    <Provider store={store}>
      <StaticRouter context={{}}>
        <_FilmList
          query="a query"
          films={films}
          totalResults={films.length}
          dispatchSetQuery={dispatchSetQuery}
        />
      </StaticRouter>
    </Provider>
  )
  expect(testInstance).toMatchSnapshot()
  expect(testInstance.root.findAllByType(FilmTitle)).toHaveLength(films.length)
})

it('if no films then calls dispatchSetQuery', () => {
  const dispatchSetQuery = jest.fn()
  create(
    <Provider store={store}>
      <StaticRouter context={{}}>
        <_FilmList
          query="a query"
          totalResults={0}
          dispatchSetQuery={dispatchSetQuery}
        />
      </StaticRouter>
    </Provider>
  )
  expect(dispatchSetQuery.mock.calls.length).toEqual(1)
})
