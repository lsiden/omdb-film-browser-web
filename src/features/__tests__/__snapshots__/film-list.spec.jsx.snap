// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmList renders list of titles 1`] = `
<div>
  <header
    style={
      Object {
        "background": "linear-gradient(to right, #000040, #3b00ff)",
        "color": "white",
        "padding": "10px",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <h1
        style={
          Object {
            "fontSize": 22,
            "marginTop": 10,
          }
        }
      />
      <a
        href="/about"
        style={
          Object {
            "color": "white",
          }
        }
      >
        about
      </a>
    </div>
    <form
      style={
        Object {
          "marginBottom": "8pt",
          "minWidth": "320px",
        }
      }
    >
      <label>
        Search
        <div
          style={
            Object {
              "display": "flex",
              "flex": "1 1",
              "maxWidth": 600,
              "position": "relative",
              "width": "100%",
            }
          }
        >
          <span
            charSet="utf-8"
            dangerouslySetInnerHTML={
              Object {
                "__html": "⚲",
              }
            }
            style={
              Object {
                "color": "#888",
                "fontSize": "20pt",
                "left": "6pt",
                "position": "absolute",
                "top": "3pt",
                "transform": "rotate(-45deg)",
              }
            }
          />
          <input
            onChange={[Function]}
            placeholder="Title"
            style={
              Object {
                "color": "black",
                "fontSize": "12pt",
                "height": "24pt",
                "marginTop": "4pt",
                "maxWidth": "640px",
                "textIndent": "24pt",
                "width": "100%",
              }
            }
            type="search"
            value=""
          />
          <span
            aria-label="Cancel"
            charSet="utf-8"
            className="hover-pointer"
            dangerouslySetInnerHTML={
              Object {
                "__html": "❌",
              }
            }
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            style={
              Object {
                "color": "#888",
                "fontSize": "18pt",
                "position": "absolute",
                "right": 5,
                "top": 5,
              }
            }
            tabIndex={0}
          />
        </div>
      </label>
    </form>
  </header>
  <div>
    <ul
      style={
        Object {
          "lineHeight": 1.5,
          "listStyleType": "none",
          "marginTop": 20,
        }
      }
    >
      <div>
        <a
          href="/film/tt2310332"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          The Hobbit: The Battle of the Five Armies, 2014
        </a>
      </div>
      <div>
        <a
          href="/film/tt1217613"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Battle Los Angeles, 2011
        </a>
      </div>
      <div>
        <a
          href="/film/tt0266308"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Battle Royale, 2000
        </a>
      </div>
      <div>
        <a
          href="/film/tt1078912"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Night at the Museum: Battle of the Smithsonian, 2009
        </a>
      </div>
      <div>
        <a
          href="/film/tt0058946"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          The Battle of Algiers, 1966
        </a>
      </div>
      <div>
        <a
          href="/film/tt4622512"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Battle of the Sexes, 2017
        </a>
      </div>
      <div>
        <a
          href="/film/tt0069768"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Battle for the Planet of the Apes, 1973
        </a>
      </div>
      <div>
        <a
          href="/film/tt2263944"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Dragon Ball Z: Battle of Gods, 2013
        </a>
      </div>
      <div>
        <a
          href="/film/tt2536562"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Dragon Ball Z: Battle of Gods, 2013
        </a>
      </div>
      <div>
        <a
          href="/film/tt0064072"
          onClick={[Function]}
          style={
            Object {
              "color": undefined,
            }
          }
          title="View film details"
        >
          Battle of Britain, 1969
        </a>
      </div>
    </ul>
    <div
      style={
        Object {
          "alignItems": "baseline",
          "display": "flex",
          "justifyContent": "flex-start",
          "marginLeft": 40,
        }
      }
    >
      <button
        className="btn btn-link"
        onClick={[Function]}
        style={
          Object {
            "color": "darkgrey",
            "fontSize": 16,
            "marginLeft": 40,
          }
        }
        type="button"
      >
        top
      </button>
    </div>
  </div>
</div>
`;
